### PR TITLE
Add transactions multi-signing.

### PR DESCRIPTION
**Pull request by dzmitryhil**
### Add transaction multi-signing.

* Add `MultiSign` and `SetSigners` methods for transactions to be signed from multi-signing.
* Add an additional `MultiSignable` interface for the `BaseTx` and implement it. 
*  Fix structure used by `SIGNER_LIST_SET`  and `multi-signing` 
*  Add missing `temBAD_WEIGHT` error code

Resolves https://github.com/rubblelabs/ripple/issues/40
Resolves https://github.com/rubblelabs/ripple/issues/44 
Resolves https://github.com/rubblelabs/ripple/issues/75

### CODE REVIEW

File: data/format.go [hunk 1]
```diff
@@ -14,14 +14,15 @@ type LedgerNamespace uint16
...
-	HP_TRANSACTION_SIGN HashPrefix = 0x53545800 // 'STX' inner transaction to sign
-	HP_VALIDATION       HashPrefix = 0x56414C00 // 'VAL' validation for signing
-	HP_PROPOSAL         HashPrefix = 0x50525000 // 'PRP' proposal for signing
+	HP_TRANSACTION_ID        HashPrefix = 0x54584E00 // 'TXN' transaction
+	HP_TRANSACTION_NODE      HashPrefix = 0x534E4400 // 'SND' transaction plus metadata (probably should have been TND!)
+	HP_LEAF_NODE             HashPrefix = 0x4D4C4E00 // 'MLN' account state
+	HP_INNER_NODE            HashPrefix = 0x4D494E00 // 'MIN' inner node in tree
+	HP_LEDGER_MASTER         HashPrefix = 0x4C575200 // 'LWR' ledger master data for signing (probably should have been LGR!)
+	HP_TRANSACTION_SIGN      HashPrefix = 0x53545800 // 'STX' inner transaction to sign
+	HP_TRANSACTION_MILTISIGN HashPrefix = 0x534D5400 // 'SMT' inner transaction to multi-sign
```
Comment (donovanhide) [inline data/format.go:23]: typo

File: data/ledgerentry.go [hunk 2]
```diff
@@ -133,12 +133,16 @@ type Escrow struct {
 	DestinationNode *NodeIndex       `json:",omitempty"`
 }
 
-type SignerEntry struct {
+type SignerEntryItem struct {
```
Comment (donovanhide) [inline data/ledgerentry.go:136]: Have you got an example tx or ledge entry from mainnet which demonstrates this structure?

File: data/memo.go [hunk 3]
```diff
@@ -1,11 +1,13 @@
 package data
 
+type MemoItem struct {
```
Comment (donovanhide) [inline data/memo.go:3]: Why the change?

File: websockets/commands.go [hunk 4]
```diff
@@ -16,9 +16,10 @@ type Syncer interface {
 }
 
 type CommandError struct {
-	Name    string `json:"error"`
-	Code    int    `json:"error_code"`
-	Message string `json:"error_message"`
+	Name      string `json:"error"`
+	Code      int    `json:"error_code"`
+	Message   string `json:"error_message"`
+	Exception string `json:"error_exception"`
```
Comment (donovanhide) [inline websockets/commands.go:22]: Is this a new response field from somewhere? Docs?

**Review by donovanhide — COMMENTED** [review]
Hi, lots of changes in a single commit. Would be good to break it into separate PR's in case you break anyone's code.

File: data/ledgerentry.go (hunk 2 above)
Comment (dzmitryhil) [reply data/ledgerentry.go:136]: * Doc: https://xrpl.org/signerlistset.html#example-signerlistset-json
* Here is the tesnet example (unfortunately I don't have the same for the mainnet): https://testnet.xrpl.org/transactions/01A6D400C0F114543A3BE26708F9E3CBBFDDDF40000E1D5C85E3F0DDFE5EC9C1/raw

File: data/format.go (hunk 1 above)
Comment (dzmitryhil) [reply data/format.go:23]: Do you mean to split to `MILTI_SIGN`?

Comment (donovanhide) [reply data/format.go:23]: s/MILTI_SIGN/MULTI_SIGN/

Comment (dzmitryhil) [reply data/format.go]: Right, fixed.

File: data/memo.go (hunk 3 above)
Comment (donovanhide) [reply data/memo.go:3]: I have been doing this:

```		memos := make(data.Memos, 1)
		memos[0].Memo.MemoData = data.VariableLength(quoteData)```
Not  particularly elegant, but does your change break this code?
Irrespective of this, it's a change being slipped into a PR for another problem, so best to break it out into a separate issue/PR.
After that happy to merge.

File: websockets/commands.go (hunk 4 above)
Comment (donovanhide) [reply websockets/commands.go:22]: This should be in a separate PR too. Just in case anyone needs to check the commit history to easily see where changes have been added.

File: data/ledgerentry.go (hunk 2 above)
Comment (dzmitryhil) [reply data/ledgerentry.go:136]: My bad, that was the proof for the  `Signer` and `SignerItem`.
And here is the proof for the `SignerListSet`: https://testnet.xrpl.org/transactions/05AADAF31482E06241870D93D32FDC5E59414D7A9D60754D1C1DEC5ACFDF3860/raw

File: websockets/commands.go (hunk 4 above)
Comment (dzmitryhil) [reply websockets/commands.go]: Removed from that PR. And will be in a different PR.

File: data/memo.go (hunk 3 above)
Comment (dzmitryhil) [reply data/memo.go]: The proposed changes might be a breaking change depending on the way the memo is initialized. If it's initialized in the way you do - not breaching, but if in the way where you provide the embedded struct, it's breaking. 

That change is removed from that PR and will be moved to another.

Comment (dzmitryhil) [reply data/memo.go]: https://github.com/rubblelabs/ripple/pull/77
